File: testbed/Manager/extensions/gui/src/lib/objects/js/select.js
import {Widget} from "../objects.js";
import {getColor, shadeColor, interpolateColors} from "../../helpers.js";

export class MultiSelectWidget extends Widget {
    constructor(id, config = {}) {
        super(id, config);

        const defaults = {
            visible: true,
            color: "#333",        // or a single color; per-option colors can be specified in options[color]
            text_color: "#fff",
            title: "",
            options: {},          // { valueKey: { label, color? }, … }
            value: null,          // the selected valueKey
            lockable: false,
            locked: false,
            titlePosition: "top",  // 'top' or 'left'
            // titleStyle: "bold"     // 'bold' or 'normal'
        };

        this.configuration = {...defaults, ...this.configuration};
        if (!this.configuration.lockable) {
            this.configuration.locked = false;
        }

        this.element = this._initializeElement();
        this.configureElement(this.element);
        this.assignListeners(this.element);
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Determine the current background‐color. If the currently selected option
     * has its own .color property, use that; otherwise fall back to config.color.
     */
    _getCurrentColor() {
        const {color, options, value} = this.configuration;
        const optObj = options[value];
        if (optObj && optObj.color) {
            return optObj.color;
        }
        return color;
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Find the label corresponding to the current `value`.
     */
    _getCurrentLabel() {
        const {options, value} = this.configuration;
        const found = options[value];
        return found ? found.label : "";
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Create the root container. Inner content is built in `configureElement()`.
     */
    _initializeElement() {
        const container = document.createElement("div");
        container.id = this.id;
        container.classList.add("widget",'highlightable', "multiSelectWidget");
        return container;
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Apply `this.configuration` to the DOM: set styles, rebuild innerHTML, repopulate <select>.
     */
    configureElement(element) {
        super.configureElement(element);
        const c = this.configuration;
        const container = this.element;

        // ── Visibility ───────────────────────────────────────────────────────────────────────────────────────────────
        if (!c.visible) {
            container.style.display = "none";
        } else {
            container.style.display = "";
        }

        // ── Dataset flags for CSS layout or external styling ─────────────────────────────────────────────────────────
        container.dataset.titlePosition = c.title_position;
        container.dataset.lockable = c.lockable;
        container.dataset.locked = c.locked;

        // ── Colors ───────────────────────────────────────────────────────────────────────────────────────────────────
        container.style.backgroundColor = getColor(this._getCurrentColor());
        container.style.color = getColor(c.text_color);

        // ── Build inner HTML ─────────────────────────────────────────────────────────────────────────────────────────
        let html = "";
        if (c.title) {
            html += `<span class="msSelectTitle">${c.title}</span>`;
        }
        html += `
            <span class="msSelectValue">${this._getCurrentLabel()}</span>
            <select></select>
            <span class="msSelectDropdown">&#x25BC;</span>
        `;
        container.innerHTML = html;

        // ── Populate <select> with options (now `options` is an object) ───────────────────────────────────────────────
        const select = container.querySelector("select");
        // Clear any existing children (in case configureElement is called again)
        select.innerHTML = "";

        Object.entries(c.options).forEach(([optValue, optDef]) => {
            const o = document.createElement("option");
            o.value = optValue;
            o.textContent = optDef.label;
            if (optValue === c.value) {
                o.selected = true;
            }
            select.appendChild(o);
        });

        select.disabled = !!c.locked;

        // ── Stretch the <select> invisibly to capture clicks ──────────────────────────────────────────────────────────
        Object.assign(select.style, {
            position: "absolute",
            top: "0",
            left: "0",
            width: "100%",
            height: "100%",
            opacity: "0",
            cursor: "pointer",
            zIndex: "1",
        });

        // ── Style the dropdown arrow ─────────────────────────────────────────────────────────────────────────────────
        const arrow = container.querySelector(".msSelectDropdown");
        Object.assign(arrow.style, {
            position: "absolute",
            bottom: "1px",
            right: "1px",
            pointerEvents: "none",
            zIndex: "2",
        });

        // ── Apply titleStyle (font‐weight) ───────────────────────────────────────────────────────────────────────────
        // const titleEl = container.querySelector(".msSelectTitle");
        // if (titleEl) {
        //     titleEl.style.fontWeight = (c.titleStyle === "bold" ? "bold" : "normal");
        // }
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    setValue(value) {
        const c = this.configuration;
        const el = this.element;
        const select = el.querySelector("select");
        const valueEl = el.querySelector(".msSelectValue");

        if (!(value in c.options)) return; // ignore unknown values

        // Update internal state
        c.value = value;

        // Update <select>
        if (select) {
            select.value = value;
        }

        // Update visible label
        if (valueEl) {
            valueEl.textContent = this._getCurrentLabel();
        }

        // Update background color if per-option color exists
        el.style.backgroundColor = getColor(this._getCurrentColor());
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Expose the root element for insertion into the DOM.
     */
    getElement() {
        return this.element;
    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Merge `data` into configuration, then re-render.
     * Reassign listeners because the inner <select> may be rebuilt.
     */
    update(data) {

    }

    /* -------------------------------------------------------------------------------------------------------------- */
    /**
     * Wire up event listeners on the <select> and the container for change, right-click (lock prevention),
     * and long-press → long-click detection.
     */
    assignListeners(container) {

        // Clear any previously attached listeners by cloning the node
        // (Prevents duplicate handlers if `update()` is called repeatedly.)
        const fresh = container.cloneNode(false);
        container.parentNode?.replaceChild(fresh, container);
        this.element = fresh;

        // Re-apply the contents (they were lost by cloneNode). Then proceed to attach listeners.
        this.configureElement(this.element);
        super.assignListeners(this.element);

        const select = this.element.querySelector("select");
        const c = this.configuration;
        const valueEl = this.element.querySelector(".msSelectValue");

        // ── On selection change ───────────────────────────────────────────────────────────────────────────────────────
        select.addEventListener("change", () => {
            c.value = select.value;
            valueEl.textContent = this._getCurrentLabel();

            this.callbacks.get('event').call({
                id: this.id,
                event: "multi_select_change",
                data: {value: c.value},
            });

            this.element.style.backgroundColor = getColor(this._getCurrentColor());
            this.element.classList.add("accepted");
            this.element.addEventListener(
                "animationend",
                () => {
                    this.element.classList.remove("accepted");
                },
                {once: true}
            );
        });

        // ── Prevent default context menu if lockable ─────────────────────────────────────────────────────────────────
        // select.addEventListener("contextmenu", (e) => {
        //     if (c.lockable) {
        //         e.preventDefault();
        //     }
        // });

        // ── Long-press detection for long-click event ──────────────────────────────────────────────────────────────
        let longPressTimer;
        const LP_THRESHOLD = 500;
        const startPress = () => {
            longPressTimer = setTimeout(() => {
                this.callbacks.get('event').call({
                    id: this.id,
                    event: "multi_select_long_click",
                    data: {},
                });
            }, LP_THRESHOLD);
        };
        const clearPress = () => {
            clearTimeout(longPressTimer);
        };

        this.element.addEventListener("mousedown", startPress);
        this.element.addEventListener("mouseup", clearPress);
        this.element.addEventListener("mouseleave", clearPress);
        this.element.addEventListener("touchstart", startPress, {passive: true});
        this.element.addEventListener("touchend", clearPress);
        this.element.addEventListener("touchcancel", clearPress);
    }

    updateConfig(data) {
        this.configuration = {...this.configuration, ...data};
        this.configureElement(this.element);
        this.assignListeners(this.element);
    }
}

File: testbed/Manager/extensions/gui/src/lib/objects/python/select.py
from typing import Any

from core.utils.callbacks import callback_definition, CallbackContainer
from core.utils.logging_utils import Logger
from extensions.gui.src.lib.objects.objects import Widget


@callback_definition
class MultiSelectWidgetCallbacks:
    """
    Defines callbacks for MultiSelectWidget:
      - selection_changed: fired when the user picks a new option
      - long_click: fired when the user long-presses on the widget
    """
    selection_changed: CallbackContainer
    long_click: CallbackContainer


class MultiSelectWidget(Widget):
    """
    A drop-down (multi-select) widget that shows a list of options, displays the current selection,
    and optionally can be locked to prevent changes.
    """

    type = "multi_select"
    callbacks: MultiSelectWidgetCallbacks

    # === INIT =========================================================================================================
    def __init__(
            self,
            widget_id: str,
            *,
            options: dict = None,
            value=None,
            config: dict | None = None,
            **kwargs
    ):
        super().__init__(widget_id)

        self.logger = Logger(f"MultiSelectWidget {self.id}", "DEBUG")
        self.callbacks = MultiSelectWidgetCallbacks()

        if config is None:
            config = {}

        # Default configuration values
        default_config = {
            "lockable": False,
            "locked": False,
            "title": None,
            "title_position": 'top',  # 'top' or 'left'
            "titleStyle": 'bold',
            "color": [0.2, 0.2, 0.2],
            "text_color": [1, 1, 1],
            "visible": True,
        }

        # Merge defaults, caller args, and any extra kwargs
        self.config = {**default_config, **config, **kwargs}

        if self.config['title'] is None:
            self.config['title'] = self.id

        self.options = options

        # Check if options is looking correct
        if not isinstance(options, dict):
            raise TypeError("options must be a dict")

        if value is not None:
            if value not in self.options:
                raise ValueError(f"Value {value} is not a valid option")

            self.value = value

        # Validate that if color is a list, its length matches options
        col = self.config["color"]
        opts = self.options
        if isinstance(col, list) and len(col) != len(opts) and not (
                isinstance(col, list) and all(isinstance(c, (str, list)) for c in col)
        ):
            # If it’s a list of colors, expect length == len(options).
            self.logger.debug("Color list length does not match number of options; ignoring per-option mapping.")

    # ------------------------------------------------------------------------------------------------------------------
    @property
    def value(self) -> Any:
        return self._value

    @value.setter
    def value(self, value: Any):
        if value not in self.options:
            self.logger.error(f"Value {value} is not a valid option")
        else:
            self._value = value
        self._sendValueToFrontend(self.value)

    # ------------------------------------------------------------------------------------------------------------------
    def getConfiguration(self) -> dict:
        """
        Returns a dictionary describing this widget’s state for front-end rendering.
        """
        config = {
            'type': 'multi_select',
            'id': self.id,
            'options': self.options,
            'value': self.value,
            **self.config,
        }

        return config

    # ------------------------------------------------------------------------------------------------------------------
    def handleEvent(self, message: dict, sender=None) -> Any:
        """
        Handle incoming messages from the front-end. Expected message shapes:
          { "event": "multi_select_change", "data": { "value": new_value } }
          { "event": "multi_select_long_click" }
        """
        self.logger.debug(f"Received message: {message}")

        if 'event' not in message:
            self.logger.warning(f"Unknown message received: {message}")
            return

        match message['event']:
            case 'multi_select_change':
                self.value = message['data']['value']

        # event = message.get("event")
        # data = message.get("data", {})
        #
        # if event == "multi_select_change":
        #     new_val = data.get("value")
        #     # Only proceed if new_val exists among options or is None
        #     valid_values = [opt["value"] for opt in self.config["options"]]
        #     if new_val not in valid_values and new_val is not None:
        #         self.logger.debug(f"Ignoring invalid selection '{new_val}'.")
        #         return
        #
        #     self.config["value"] = new_val
        #     self.logger.debug(f"Selection changed to {new_val!r}")
        #
        #     # Fire the callback
        #     for cb in self.callbacks.selection_changed:
        #         cb(widget=self, value=new_val)
        #
        # elif event == "multi_select_long_click":
        #     self.logger.debug("Long-click detected on MultiSelectWidget")
        #     for cb in self.callbacks.long_click:
        #         cb(widget=self)
        #
        # return None

    # ------------------------------------------------------------------------------------------------------------------
    def _sendValueToFrontend(self, value):
        self.function(
            function_name='setValue',
            args=value
        )

    # ------------------------------------------------------------------------------------------------------------------
    # ------------------------------------------------------------------------------------------------------------------
    # ------------------------------------------------------------------------------------------------------------------
    def init(self, *args, **kwargs):
        """
        Called once when the GUI is first constructed.
        You may push the initial state to the front-end here if needed.
        """
        pass
